### DISCUSSION TITLE
Purpose of Promise-wrapper in spy mock

### DISCUSSION THREAD

**Discussion by Namchee** [vitest-dev/vitest#3824] [General]
I've noticed that the function `mockImplementation` in spy instances accepts a function that returns a Promise-wrapped value of the original type.

`mockImplementation(fn: ((...args: TArgs) => TReturns) | (() => Promise<TReturns>)): this`

What's the motivation of allowing Promise-wrapped function here? This can possibly lead to invalid mocking, for example:

```ts
// util.ts
export function addNumber(a: number, b: number): number {
  return a + b;
}

// ...
// main.ts
import { addNumber } from './util';
function sampleFunction() {
  addNumber(1 + 2);
}

// ...
// main.test.ts
import { vi } from 'vitest';

import * as util from './util';

describe('sampleFunction', () => {
  test('do something', () => {
    // this compiles, but will lead to incorrect behavior
    const spy = vi.spyOn(util, 'addNumber').mockImplementation(() => Promise.resolve(2));
  });
});
```

Related to #3750

**Comment by sheremet-va** [comment]
Yeah, there is an open pr that fixes it

**Reply by rodlukas** [reply]
@sheremet-va is there any link to the PR please? we're facing some type issues that are similar to the problem described here but we're not able to find any PR that could fix that

**Reply by sheremet-va** [reply]
I think the comment was about `withImplementation` at the time. `mockImplementation` should be fixed now with #4859
